pkg/internal/apply/layerApplier.go

```diff
@@ -151,6 +149,55 @@ func (a KubectlLayerApplier) decodeAddons(layer layers.Layer,
 	}
 }
 
+func (a KubectlLayerApplier) addOwnerRefs(layer layers.Layer, hrs []*helmopv1.HelmRelease) error {
+	for i, hr := range hrs {
+		a.logDebug("Adding owner ref to HelmRelease for AddonsLayer", layer, "index", i, "helmRelease", hr)
+		err := controllerutil.SetControllerReference(layer.GetAddonsLayer(), hr, a.scheme)
+		if err != nil {
+			// could not apply owner ref for object
+			return fmt.Errorf("unable to apply owner reference for AddonsLayer '%s' to HelmRelease '%s': %w", layer.GetName(), getLabel(hr), err)
+		}
+	}
+	return nil
+}
+
+func (a KubectlLayerApplier) getHelmRelease(hr *helmopv1.HelmRelease) (*helmopv1.HelmRelease, error) {
+	key, err := client.ObjectKeyFromObject(hr)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get an ObjectKey from HelmRelease '%s'", getLabel(hr))
+	}
+	foundHr := &helmopv1.HelmRelease{}
+	err = a.client.Get(context.Background(), key, foundHr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to Get HelmRelease '%s'", getLabel(hr))
+	}
+	return foundHr, nil
+}
+
+func (a KubectlLayerApplier) applyHelmReleases(layer layers.Layer, hrs []*helmopv1.HelmRelease) error {
+	for i, hr := range hrs {
+		a.logDebug("Applying HelmRelease for AddonsLayer", layer, "index", i, "helmRelease", hr)
+		foundHr, err := a.getHelmRelease(hr)
+		if err != nil {
+			// TODO - not sure if logic goes here if the HelmRelease does not exist on the cluster
+
+		} else if foundHr == nil {
+			// HelmRelease does not exist, create resource
+			err = a.client.Create(context.Background(), hr, &client.CreateOptions{})
+			if err != nil {
+				return fmt.Errorf("unable to Create HelmRelease '%s' on the target cluster", getLabel(hr))
+			}
+		} else {
+			// HelmRelease exists, update resource
+			err = a.client.Update(context.Background(), hr, &client.UpdateOptions{})
+			if err != nil {
+				return fmt.Errorf("unable to Update HelmRelease '%s' on the target cluster", getLabel(hr))
+			}
+		}
+	}
+	return nil
+}
+
 func (a KubectlLayerApplier) decodeList(layer layers.Layer,
 	raws *corev1.List, dez *runtime.Decoder) (hrs []*helmopv1.HelmRelease, errz []error, err error) {
```

Comment:
Returning `errz []error, err error`...this could be simplified.

Reply:
Future optimization.